src/Client/HedgedConnectionsFactory.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <Common/TimerDescriptor.h>
+#include <Client/ConnectionPoolWithFailover.h>
+#include <Core/Settings.h>
+#include <Common/Epoll.h>
+#include <unordered_map>
+#include <memory>
+
+namespace DB
+{
+
+enum class ConnectionTimeoutType
+{
+    RECEIVE_HELLO_TIMEOUT,
+    RECEIVE_TABLES_STATUS_TIMEOUT,
+    RECEIVE_DATA_TIMEOUT,
+    RECEIVE_TIMEOUT,
+};
+
+struct ConnectionTimeoutDescriptor
+{
+    ConnectionTimeoutType type;
+    TimerDescriptor timer;
+};
+
+using ConnectionTimeoutDescriptorPtr = std::shared_ptr<ConnectionTimeoutDescriptor>;
+using TimerDescriptorPtr = std::shared_ptr<TimerDescriptor>;
+
+/** Class for establishing hedged connections with replicas.
+  * The process of establishing connection is divided on stages, on each stage if
+  * replica doesn't respond for a long time, we start establishing connection with
+  * the next replica, without cancelling working with previous one.
+  * It works with multiple replicas simultaneously without blocking by using epoll.
+  */
+class HedgedConnectionsFactory
+{
+public:
+    using ShuffledPool = ConnectionPoolWithFailover::Base::ShuffledPool;
+
+    enum class State
+    {
+        EMPTY = 0,
```

Comment:
Note that EMPTY is an internal state. Maybe it is ok to have it here, but maybe we an remove it...